### ISSUE TITLE
NER with Prompting

### ISSUE THREAD

**Issue by yigittuncer07** [yigittuncer07/nlp#21]

**Comment by yigittuncer07** [comment]
Here is some code that CONSUMES from the TGI instance (grammar and all)

``` python
from huggingface_hub import InferenceClient
from datasets import load_dataset
from sklearn.metrics import precision_recall_fscore_support, accuracy_score, confusion_matrix
import json
import logging
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime

logging.basicConfig(
    filename='ner_inference-1b.log',
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)

PROMPT = "Aşağıdaki kelime listesi üzerinde Adlandırılmış Varlık Tanıma (NER) işlemi yap ve her bir kelimeyi şu etiketlerden biriyle sınıflandır: B-LOC (konumun başlangıcı), I-LOC (konumun içi), B-ORG (kuruluşun başlangıcı), I-ORG (kuruluşun içi), B-PER (kişinin başlangıcı), I-PER (kişinin içi), O (diğer).\n\nKelime Listesi: '{}'\n\nSonucu 'entities' adlı bir dizi içeren bir JSON nesnesi olarak döndür. Her varlık 'kelime' ve 'etiket' özelliklerine sahip olmalıdır."
DATASET_NAME = "turkishnlp/WikiANN-Turkish-JSON-Format"

dataset = load_dataset(DATASET_NAME)
dataset = dataset['validation']

client = InferenceClient("http://localhost:1410")

ner_schema = {
    "type": "object",
    "properties": {
        "entities": {
            "type": "array",
            "items": {
                "type": "object",
                "properties": {
                    "kelime": {"type": "string"},
                    "etiket": {"type": "string", "enum": ["B-LOC", "B-ORG", "B-PER", "I-LOC", "I-ORG", "I-PER", "O"]}
                },
                "required": ["kelime", "etiket"]
            }
        }
    },
    "required": ["entities"]
}

def ner_infer(words):
    try:
        resp = client.text_generation(
            PROMPT.format(words),
            max_new_tokens=512,
            seed=42,
            grammar={"type": "json", "value": ner_schema},
        )
        return resp
    except Exception as e:
        logging.error(f"Error during inference: {e}")
        return None

def evaluate_ner(dataset, num_samples=None):
    if num_samples:
        subset = dataset[:num_samples]
    else:
        subset = dataset
        
    tokens_list = subset["tokens"]
    tags_list = subset["tags"]
    
    all_true_tags = []
    all_pred_tags = []
    broken_responses_count = 0
    
    for idx, (tokens, tags) in enumerate(zip(tokens_list, tags_list)):
        logging.info(f"Processing sample {idx+1}/{len(tokens_list)}")
        
        response = ner_infer(str(tokens))
        
        logging.info(f"Sample {idx+1} - Raw response: {response}")
        
        try:
            result = json.loads(response)
            entities = result.get("entities", [])
            
            if len(entities) != len(tokens):
                logging.warning(f"Sample {idx+1} - Length mismatch: entities({len(entities)}) != tokens({len(tokens)})")
                broken_responses_count += 1
                continue
                
            pred_tags = [entity["etiket"] for entity in entities]
            
            all_true_tags.extend(tags)
            all_pred_tags.extend(pred_tags)
            
        except json.JSONDecodeError:
            logging.error(f"Sample {idx+1} - Invalid JSON response")
            broken_responses_count += 1
            continue
        except Exception as e:
            logging.error(f"Sample {idx+1} - Error processing response: {e}")
            broken_responses_count += 1
            continue
    
    if len(all_true_tags) > 0 and len(all_pred_tags) > 0:
        tags_set = sorted(list(set(all_true_tags + all_pred_tags)))
        tag_to_idx = {tag: idx for idx, tag in enumerate(tags_set)}
        
        true_indices = [tag_to_idx[tag] for tag in all_true_tags]
        pred_indices = [tag_to_idx[tag] for tag in all_pred_tags]
        
        precision, recall, f1, _ = precision_recall_fscore_support(
            true_indices, pred_indices, average='weighted'
        )
        macro_precision, macro_recall, macro_f1, _ = precision_recall_fscore_support(
            true_indices, pred_indices, average='macro'
        )
        accuracy = accuracy_score(true_indices, pred_indices)
        conf_matrix = confusion_matrix(true_indices, pred_indices)
        
        results = {
            "total_samples": len(tokens_list),
            "broken_responses": broken_responses_count,
            "valid_samples": len(tokens_list) - broken_responses_count,
            "metrics": {
                "accuracy": float(accuracy),
                "weighted": {
                    "precision": float(precision),
                    "recall": float(recall),
                    "f1": float(f1)
                },
                "macro": {
                    "precision": float(macro_precision),
                    "recall": float(macro_recall),
                    "f1": float(macro_f1)
                }
            },
            "confusion_matrix": conf_matrix.tolist(),
            "labels": tags_set
        }
        
        return results
    else:
        logging.error("No valid predictions collected, cannot calculate metrics")
        return None

def save_results_to_markdown(results):
    if results is None:
        return
    
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    filename = f"ner_evaluation_results_{timestamp}.md"
    
    with open(filename, "w") as f:
        f.write("# Named Entity Recognition Evaluation Results\n\n")
        f.write(f"Date: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n\n")
        
        f.write("## Dataset Information\n")
        f.write(f"- Dataset: {DATASET_NAME}\n")
        f.write(f"- Total samples evaluated: {results['total_samples']}\n")
        f.write(f"- Broken responses: {results['broken_responses']} ({results['broken_responses']/results['total_samples']*100:.2f}%)\n")
        f.write(f"- Valid samples: {results['valid_samples']} ({results['valid_samples']/results['total_samples']*100:.2f}%)\n\n")
        
        f.write("## Performance Metrics\n")
        f.write(f"- Accuracy: {results['metrics']['accuracy']:.4f}\n\n")
        
        f.write("### Weighted Metrics\n")
        f.write(f"- Precision: {results['metrics']['weighted']['precision']:.4f}\n")
        f.write(f"- Recall: {results['metrics']['weighted']['recall']:.4f}\n")
        f.write(f"- F1 Score: {results['metrics']['weighted']['f1']:.4f}\n\n")
        
        f.write("### Macro Metrics\n")
        f.write(f"- Precision: {results['metrics']['macro']['precision']:.4f}\n")
        f.write(f"- Recall: {results['metrics']['macro']['recall']:.4f}\n")
        f.write(f"- F1 Score: {results['metrics']['macro']['f1']:.4f}\n\n")
        
        f.write("## Confusion Matrix\n")
        f.write("```\n")
        f.write(f"Labels: {results['labels']}\n\n")
        
        conf_matrix = np.array(results['confusion_matrix'])
        formatted_matrix = ""
        for row in conf_matrix:
            formatted_matrix += " ".join([f"{val:4d}" for val in row]) + "\n"
        f.write(formatted_matrix)
        f.write("```\n\n")
        
        # Save visualization of confusion matrix
        plt.figure(figsize=(10, 8))
        sns.heatmap(conf_matrix, annot=True, fmt='d', cmap='Blues',
                   xticklabels=results['labels'], yticklabels=results['labels'])
        plt.title('Confusion Matrix')
        plt.xlabel('Predicted Labels')
        plt.ylabel('True Labels')
        plt.tight_layout()
        
        # Save confusion matrix plot
        cm_filename = f"confusion_matrix_{timestamp}.png"
        plt.savefig(cm_filename)
        f.write(f"![Confusion Matrix]({cm_filename})\n")
        
    print(f"Results saved to {filename}")

# Main execution
if __name__ == "__main__":
    num_samples = 5000  # set to none for full eval
    
    print(f"Starting NER evaluation on {DATASET_NAME}...")
    results = evaluate_ner(dataset, num_samples)
    
    if results:
        save_results_to_markdown(results)
        print(f"Evaluation completed. Total samples: {results['total_samples']}, Valid: {results['valid_samples']}, Broken: {results['broken_responses']}")
        print(f"Accuracy: {results['metrics']['accuracy']:.4f}, F1-score: {results['metrics']['weighted']['f1']:.4f}")
    else:
        print("Evaluation failed due to insufficient valid responses.")
```

**Comment by yigittuncer07** [comment]
docker compose for actually launching the TGI instance:
``` bash
services:
  tgi:
    image: ghcr.io/huggingface/text-generation-inference:3.0.1
    command:
      - "--model-id meta-llama/Llama-3.2-1B-Instruct"
      
      - "--num-shard 1"
      
      - "--max-input-tokens 3000"
      - "--max-total-tokens 4000"
      - "--max-batch-prefill-tokens 3000"
      - "--max-batch-total-tokens 4000"
      - "--max-concurrent-requests 1"           

    shm_size: '1g'
    volumes:
      - /home/yigittuncer/.cache/huggingface/hub:/data/hub
    ports:
      - "1410:80"
    environment:
      - HF_TOKEN=${HF_TOKEN}
      - USE_PREFIX_CACHING=0
      - PREFIX_CACHING=0
    restart: unless-stopped
    runtime: nvidia
    deploy:
      resources:
        reservations:
          devices:
            - driver: nvidia
              device_ids: [ '3' ]
              capabilities: [ gpu ]

```

**Comment by yigittuncer07** [comment]
# Named Entity Recognition Evaluation Results 1B LLAMA INSTRUCT

Date: 2025-03-21 02:17:19

## Dataset Information
- Dataset: turkishnlp/WikiANN-Turkish-JSON-Format
- Total samples evaluated: 5000
- Broken responses: 3774 (75.48%)
- Valid samples: 1226 (24.52%)

## Performance Metrics
- Accuracy: 0.1226

### Weighted Metrics
- Precision: 0.5694
- Recall: 0.1226
- F1 Score: 0.1355

### Macro Metrics
- Precision: 0.1864
- Recall: 0.1789
- F1 Score: 0.0769

## Confusion Matrix
```
Labels: ['B-LOC', 'B-ORG', 'B-PER', 'I-LOC', 'I-ORG', 'I-PER', 'O']

 232    9    2  160    3    0    3
 241   14    1  227    8    1    3
 320   36    0  137    9    4    2
 151   12    0  168   10    0    1
 374   77   15  440   52    7   13
 172   26    5  399   38    1    4
1778  229   86 1562  159   97  486
```

**Comment by yigittuncer07** [comment]
# Named Entity Recognition Evaluation Results 3B LLAMA INSTRUCT

Date: 2025-03-21 04:07:51

## Dataset Information
- Dataset: turkishnlp/WikiANN-Turkish-JSON-Format
- Total samples evaluated: 5000
- Broken responses: 3471 (69.42%)
- Valid samples: 1529 (30.58%)

## Performance Metrics
- Accuracy: 0.1582

### Weighted Metrics
- Precision: 0.5413
- Recall: 0.1582
- F1 Score: 0.1178

### Macro Metrics
- Precision: 0.2650
- Recall: 0.2964
- F1 Score: 0.1711

## Confusion Matrix
```
Labels: ['B-LOC', 'B-ORG', 'B-PER', 'I-LOC', 'I-ORG', 'I-PER', 'O']

 299   92   52   88    9   18    0
 276  164  102   75   11   21    2
  61   30  522   36    2   14    1
 200   61   68  170   38   27    2
 374  291  182  289   70   95    8
 191   96  442  113   13  112    0
1085  289  346 1905  282  671  159
```

**Comment by yigittuncer07** [comment]
Get 1 and 5 shot results here as well

**Comment by yigittuncer07** [comment]
Instruct FT these models and get few and one-shot evaluate again
